source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -127,19 +148,24 @@ void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
     return;
   }
 
-  // https://www.rfc-editor.org/rfc/pdfrfc/rfc7234.txt.pdf
-  // 4.3.4 Freshening Stored Responses upon Validation
-  // use other header fields provided in the 304 (Not Modified)
-  // response to replace all instances of the corresponding header
-  // fields in the stored response.
-  //
   // Assumptions:
   // 1. The internet is fast, i.e. we get the result as soon as the server sends it.
-  // Race conditions would not be possible because we are always processing up-to-date data.
+  //    Race conditions would not be possible because we are always processing up-to-date data.
   // 2. No key collision for etag. Therefore, if etag matches it's the same resource.
   // 3. Backend is correct. etag is being used as a unique identifier to the resource
-  // TODO(tangsaidi) merge the header map instead of replacing it according to rfc7234
-  entry.response_headers_ = Http::createHeaderMap<Http::ResponseHeaderMapImpl>(response_headers);
+
+  // use other header fields provided in the new response to replace all instances
+  // of the corresponding header fields in the stored response
+  response_headers.iterate(
+      [&entry](const Http::HeaderEntry& response_header) -> Http::HeaderMap::Iterate {
+        Http::LowerCaseString lower_case_key{response_header.key().getStringView()};
+        if (headersNotToUpdate.contains(lower_case_key)) {
```

Comment:
This check should be moved to CacheFilter::shouldUpdateCachedEntry, which should return false if a 304 contains a Content-Length field (unless it's equal to the current one). I don't regard that as blocking this PR though, since it's a pre-existing issue.

Reply:
This basically handles individual keys that we do not want to update, given that we already decided we want to update the entry. I'm adding additional checks for these items in `shouldUpdateCachedEntry` in my next PR. I'll make sure to add you as a reviewer there.